File: red/test/storage.test.ts
import { afterEach, describe, expect, test } from "bun:test";
import { existsSync, rmSync } from "node:fs";
import { join } from "node:path";
import { runtime, type ExecOptions, type ExecResult } from "red/runtime";
import { PRESERVE_JINJA_DELIMITERS } from "red/scaffold";
import * as storage from "../src/storage.ts";
import { awsFixture, awsOptout, credentials, fixture, tempWorkdir } from "./fixtures.ts";

const originalExec = runtime.exec;
afterEach(() => { runtime.exec = originalExec; });

// A tofu and aws CLI that answer the way a fresh account would: no state, no
// bucket, an apply that mints the pair.
function cli(overrides: Partial<Record<string, ExecResult>> = {}) {
  const calls: Array<[string[], ExecOptions]> = [];
  runtime.exec = async (args: string[], options: ExecOptions = {}): Promise<ExecResult> => {
    calls.push([args, options]);
    const key = args.slice(0, 3).join(" ");
    if (overrides[key]) return overrides[key]!;
    if (key === "tofu state list") return { exit: 1, out: "", err: "Error: No state file was found!" };
    if (key === "tofu show -json") return { exit: 0, out: '{"values":{"root_module":{"resources":[]}}}', err: "" };
    if (key === "aws s3api head-bucket") return { exit: 254, out: "", err: "An error occurred (404) when calling the HeadBucket operation: Not Found" };
    if (key === "tofu output -json") return { exit: 0, out: '{"credentials":{"sensitive":true,"value":{"backup":{"access_key_id":"AKIA","secret_access_key":"s"}}}}', err: "" };
    return { exit: 0, out: "", err: "" };
  };
  return calls;
}

describe("storage", () => {
  test("the managed gate is the one key", () => {
    expect(storage.managed(awsFixture())).toBe(true);
    expect(storage.managed(awsOptout())).toBe(false);
    expect(storage.managed(fixture())).toBe(false);
    expect(storage.managed(fixture({ "redis-storage-managed": "true" }))).toBe(false);
  });

  test("the stage renders one template and no credential", () => {
    const opts = awsFixture({ "red/event": "build", [storage.credentialsKey]: credentials });
    const specs = storage.specs(opts);
    expect(specs.length).toBe(1);
    const [spec] = specs;
    expect(String(spec!.target).endsWith("/redis-aws-fixture/redis-storage/main.tf")).toBe(true);
    expect(spec!.template?.name).toBe("storage/main.tf");
    expect((spec!.data ?? {})[storage.credentialsKey]).toBeUndefined();
    expect(spec!.opts).toBe(PRESERVE_JINJA_DELIMITERS);
  });

  test("the credential env carries the operator names", () => {
    expect(storage.credentialEnv({ [storage.credentialsKey]: { credentials: { backup: { access_key_id: "fixture-id", secret_access_key: "fixture-secret" } } } }))
      .toEqual({ COLORS_PAR_REDIS_BACKUP_R2_ACCESS_KEY_ID: "fixture-id", COLORS_PAR_REDIS_BACKUP_R2_SECRET_ACCESS_KEY: "fixture-secret" });
    // A missing or blank pair is refused, never an empty variable.
    expect(() => storage.credentialEnv({})).toThrow("managed storage credentials unavailable");
    expect(() => storage.credentialEnv({ [storage.credentialsKey]: { credentials: { backup: { access_key_id: "", secret_access_key: "s" } } } })).toThrow();
  });

  test("aws env overlays only what is set", () => {
    expect(storage.awsEnv(awsFixture())).toEqual({});
    expect(storage.awsEnv(awsFixture({ "aws-access-key-id": "AKIA", "aws-secret-access-key": "s", "aws-session-token": "t" })))
      .toEqual({ AWS_ACCESS_KEY_ID: "AKIA", AWS_SECRET_ACCESS_KEY: "s", AWS_SESSION_TOKEN: "t" });
    expect(storage.awsEnv({ "aws-access-key-id": "AKIA", "aws-secret-access-key": "" })).toEqual({ AWS_ACCESS_KEY_ID: "AKIA" });
  });

  test("the step is a no-op unless managed", async () => {
    runtime.exec = async () => { throw new Error("must not run"); };
    expect((await storage.step(awsOptout({ "red/event": "create" })))["red/exit"]).toBe(0);
    expect(await storage.readCredentials(awsOptout())).toEqual(awsOptout());
  });

  test("a create runs the preflight then tofu with the aws environment", async () => {
    const workdir = tempWorkdir();
    try {
      const calls = cli();
      const opts = awsFixture({ "red/event": "create", "aws-access-key-id": "AKIA", "aws-secret-access-key": "s", workdir });
      const result = await storage.step(opts);
      expect(result["red/exit"]).toBe(0);
      expect(calls.map(([args]) => args.slice(0, 2).join(" "))).toEqual([
        "tofu init", "tofu state", "tofu show", "aws s3api", "tofu init", "tofu apply", "tofu output",
      ]);
      for (const [, options] of calls) {
        expect(options.env).toEqual({ AWS_ACCESS_KEY_ID: "AKIA", AWS_SECRET_ACCESS_KEY: "s" });
        expect(options.cwd).toBe(join(workdir, "redis-aws-fixture", "redis-storage"));
      }
      expect(result[storage.credentialsKey]).toEqual({ credentials: { backup: { access_key_id: "AKIA", secret_access_key: "s" } } });
      expect(existsSync(join(workdir, "redis-aws-fixture", "redis-storage", "main.tf"))).toBe(true);
    } finally {
      rmSync(workdir, { recursive: true, force: true });
    }
  });

  test("a delete runs no preflight and removes the rendered stage", async () => {
    const workdir = tempWorkdir();
    try {
      const calls = cli();
      const result = await storage.step(awsFixture({ "red/event": "delete", workdir }));
      expect(result["red/exit"]).toBe(0);
      expect(calls.map(([args]) => args.slice(0, 2).join(" "))).toEqual(["tofu init", "tofu destroy"]);
      expect(existsSync(join(workdir, "redis-aws-fixture", "redis-storage", "main.tf"))).toBe(false);
    } finally {
      rmSync(workdir, { recursive: true, force: true });
    }
  });

  test("an owned address skips the bucket probe", async () => {
    const workdir = tempWorkdir();
    try {
      const calls = cli({
        "tofu state list": { exit: 0, out: 'aws_s3_bucket.application["backup"]\n', err: "" },
        "tofu show -json": { exit: 0, out: '{"values":{"root_module":{"resources":[{"address":"aws_s3_bucket.application[\\"backup\\"]","values":{"bucket":"redis-aws-fixture-backup"}}]}}}', err: "" },
      });
      expect((await storage.step(awsFixture({ "red/event": "create", workdir })))["red/exit"]).toBe(0);
      expect(calls.some(([args]) => args[0] === "aws")).toBe(false);
    } finally {
      rmSync(workdir, { recursive: true, force: true });
    }
  });

  test("a refused preflight is an error with no credential in the message", async () => {
    const workdir = tempWorkdir();
    try {
      // An existing bucket the stage does not own, a 403, and a broken state
      // all fail closed.
      for (const overrides of [
        { "aws s3api head-bucket": { exit: 0, out: "", err: "" } },
        { "aws s3api head-bucket": { exit: 254, out: "", err: "An error occurred (403) when calling the HeadBucket operation: Forbidden AKIAEXAMPLE" } },
        { "tofu state list": { exit: 1, out: "", err: "AKIAEXAMPLE leaked" } },
      ]) {
        cli(overrides);
        const r = await storage.step(awsFixture({ "red/event": "create", workdir }));
        expect(r["red/exit"]).toBe(1);
        expect(String(r["red/err"])).not.toContain("AKIA");
        expect(r["red/err"]).toBe("managed S3 storage failed; inspect bucket ownership, state access, and AWS permissions");
      }
    } finally {
      rmSync(workdir, { recursive: true, force: true });
    }
  });

  test("read-credentials reads the pair back and refuses a missing output", async () => {
    const workdir = tempWorkdir();
    try {
      const calls = cli();
      const read = await storage.readCredentials(awsFixture({ "red/event": "rehearse", workdir }));
      expect(storage.credentialEnv(read).COLORS_PAR_REDIS_BACKUP_R2_ACCESS_KEY_ID).toBe("AKIA");
      expect(calls.map(([args]) => args.slice(0, 2).join(" "))).toEqual(["tofu init", "tofu output"]);
      expect(existsSync(join(workdir, "redis-aws-fixture", "redis-storage", "backend.tf.json"))).toBe(true);
      cli({ "tofu output -json": { exit: 0, out: '{"credentials":{"sensitive":true,"value":{}}}', err: "" } });
      await expect(storage.readCredentials(awsFixture({ "red/event": "rehearse", workdir })))
        .rejects.toThrow(/converge storage before rehearsal/);
    } finally {
      rmSync(workdir, { recursive: true, force: true });
    }
  });
});

File: green/test/clj/io/github/getcolors/redis/storage_test.clj
(ns io.github.getcolors.redis.storage-test
  (:require [clojure.string :as str]
            [clojure.test :refer [deftest is testing]]
            [green.scaffold :as sc]
            [green.tofu :as tofu]
            [io.github.getcolors.redis.storage :as storage]
            [io.github.getcolors.redis.validate-test :refer [fixture aws-fixture aws-optout]]))

(deftest the-managed-gate-is-the-one-key
  (is (storage/managed? (aws-fixture)))
  (is (not (storage/managed? (aws-optout))))
  (is (not (storage/managed? (fixture))))
  (is (not (storage/managed? (assoc (fixture) :redis-storage-managed "true")))))

(deftest the-stage-renders-one-template-and-no-credential
  (let [opts (assoc (aws-fixture) :green/event :build storage/credentials-key {:credentials {"backup" {"access_key_id" "AKIA" "secret_access_key" "s"}}})
        [spec :as specs] (storage/specs opts)]
    (is (= 1 (count specs)))
    (is (str/ends-with? (str (:target spec)) "/redis-aws-fixture/redis-storage/main.tf"))
    (is (= :io.github.getcolors.redis.tools.storage/main.tf (:template spec)))
    (is (nil? (get (:data spec) storage/credentials-key)))
    (is (= sc/preserve-jinja-delimiters (:opts spec)))))

(deftest the-credential-env-carries-the-operator-names
  (let [env (storage/credential-env {storage/credentials-key {:credentials {"backup" {"access_key_id" "fixture-id" "secret_access_key" "fixture-secret"}}}})]
    (is (= {"COLORS_PAR_REDIS_BACKUP_R2_ACCESS_KEY_ID" "fixture-id"
            "COLORS_PAR_REDIS_BACKUP_R2_SECRET_ACCESS_KEY" "fixture-secret"} env)))
  (testing "keyword keys from a parsed output work too"
    (is (= "k" (get (storage/credential-env {storage/credentials-key {:credentials {:backup {:access_key_id "k" :secret_access_key "s"}}}})
                    "COLORS_PAR_REDIS_BACKUP_R2_ACCESS_KEY_ID"))))
  (testing "a missing or blank pair is refused, never an empty variable"
    (is (thrown? Exception (storage/credential-env {})))
    (is (thrown? Exception (storage/credential-env {storage/credentials-key {:credentials {"backup" {"access_key_id" "" "secret_access_key" "s"}}}})))))

(deftest aws-env-overlays-only-what-is-set
  (is (= {} (storage/aws-env (aws-fixture))))
  (is (= {"AWS_ACCESS_KEY_ID" "AKIA" "AWS_SECRET_ACCESS_KEY" "s" "AWS_SESSION_TOKEN" "t"}
         (storage/aws-env (assoc (aws-fixture) :aws-access-key-id "AKIA" :aws-secret-access-key "s" :aws-session-token "t"))))
  (is (= {"AWS_ACCESS_KEY_ID" "AKIA"} (storage/aws-env {:aws-access-key-id "AKIA" :aws-secret-access-key ""}))))

(deftest the-step-is-a-no-op-unless-managed
  (let [opts (assoc (aws-optout) :green/event :create)]
    (with-redefs [tofu/tofu-with-spec (fn [& _] (throw (Exception. "must not run")))]
      (is (= 0 (:green/exit (storage/step opts))))))
  (is (= (aws-optout) (storage/read-credentials! (aws-optout)))))

(deftest an-empty-state-is-recognised-across-opentofu-releases
  (testing "OpenTofu 1.11.5 (the neon-multi-node-aws build)"
    (is (storage/empty-state? {:exit 1 :err "No state file was found!\nState management commands require a state file. Run this command\n"})))
  (testing "OpenTofu 1.12.5 (the redis-aws build), no bang and an Error: prefix"
    (is (storage/empty-state? {:exit 1 :err "\nError: No state file was found\n\nState management commands require a state file. Run this command in a\n"})))
  (testing "anything else fails closed: a failed read never means absence"
    (is (not (storage/empty-state? {:exit 0 :err "" :out ""})))
    (is (not (storage/empty-state? {:exit 1 :err "Error: error loading the remote state: AccessDenied\n"})))
    (is (not (storage/empty-state? {:exit 2 :err "No state file was found!"})))))

(defn- temp-workdir []
  (str (java.nio.file.Files/createTempDirectory "redis-test-" (make-array java.nio.file.attribute.FileAttribute 0))))

(deftest a-create-runs-the-preflight-then-tofu-with-the-aws-environment
  (let [calls (atom [])
        opts (aws-fixture :green/event :create :aws-access-key-id "AKIA" :aws-secret-access-key "s" :workdir (temp-workdir))]
    (with-redefs [storage/ownership-preflight! (fn [_] (swap! calls conj :preflight))
                  tofu/tofu-with-spec (fn [opts specs config]
                                        (swap! calls conj :tofu)
                                        (is (= 1 (count specs)))
                                        (is (= {"AWS_ACCESS_KEY_ID" "AKIA" "AWS_SECRET_ACCESS_KEY" "s"} (:env config)))
                                        (is (= storage/credentials-key (:output-key config)))
                                        (assoc opts :green/exit 0))]
      (is (= 0 (:green/exit (storage/step opts))))
      (is (= [:preflight :tofu] @calls))))
  (testing "a delete runs no preflight"
    (let [calls (atom [])]
      (with-redefs [storage/ownership-preflight! (fn [_] (swap! calls conj :preflight))
                    tofu/tofu-with-spec (fn [opts _ _] (swap! calls conj :tofu) (assoc opts :green/exit 0))]
        (is (= 0 (:green/exit (storage/step (aws-fixture :green/event :delete :workdir (temp-workdir))))))
        (is (= [:tofu] @calls)))))
  (testing "a refused preflight is an error with no credential in the message"
    (with-redefs [storage/ownership-preflight! (fn [_] (throw (ex-info "AKIAEXAMPLE leaked" {})))]
      (let [r (storage/step (aws-fixture :green/event :create :workdir (temp-workdir)))]
        (is (= 1 (:green/exit r)))
        (is (not (str/includes? (:green/err r) "AKIA")))))))

(deftest read-credentials-refuses-a-missing-output
  (with-redefs-fn {#'tofu/outputs (fn [& _] {:credentials {}})
                   #'storage/checked (fn [& _] "")}
    (fn []
      (is (thrown-with-msg? Exception #"converge storage before rehearsal"
                            (storage/read-credentials! (aws-fixture :green/event :rehearse :workdir (temp-workdir))))))))
